### PR TITLE
Add Deep Evidential Regression

### LINKED ISSUE #32: Add Deep Evidential Regression
Now that we support regression, it could be interesting to implement simple methods, such as [Deep Evidential Regression ](https://proceedings.neurips.cc/paper/2020/file/aab085461de182608ee9f607f3f7d18f-Paper.pdf) (DER).

This method would be relatively straightforward to implement. Here is a possible roadmap:

- [ ] Add Deep Evidential Regression's loss
- [ ] Change "dist_estimation" parameter of the regression routine by an integer parameter to choose the number of parameters that we have to predict for the distribution
- [ ] Add tests
- [ ] Add a tutorial to explain how to use DER
- [ ] Add the reference to the documentation

### PR DESCRIPTION
**Pull request by badrmarani**
This PR fixes #32. The main changes are:
- [x] Add Deep Evidential Regression's loss.
- [x] Change the `dist_estimation` parameter of the regression routine from a boolean to an integer in order to specify the number of parameters to estimate.
- [x] Add tests.
- [x] Add a tutorial to explain how to use DER.
- [x] Add the reference to the documentation.
- [x] Add a synthetic toy dataset.

### CODE REVIEW

File: torch_uncertainty/datamodules/cubic_regression.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
+from argparse import ArgumentParser
+from typing import Any, Optional
+
+import torch
+from pytorch_lightning import LightningDataModule
+from torch.utils.data import DataLoader, Dataset, TensorDataset, random_split
+
+
+class CubicDataModule(LightningDataModule):
```
Comment (o-laurent) [inline torch_uncertainty/datamodules/cubic_regression.py:9]: I'll add some docstrings here

File: torch_uncertainty/datamodules/cubic_regression.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
+from argparse import ArgumentParser
```
Comment (o-laurent) [inline torch_uncertainty/datamodules/cubic_regression.py:1]: If we keep the datamodule, it may be preferable to put it into a "toy" folder or file, for instance. However, I'm not entirely sure it is relevant to include it in the library, at least here. @alafage any thoughts? We could also put it into the utils folder? Another possibility is to implement it as a dataset, which makes more sense as we expect much more diversity on this side.

Depending on what we make of the datamodule, we could generalize it to polynomial expressions.

File: tests/test_losses.py [hunk 3]
```diff
@@ -34,3 +34,14 @@ def test_no_bayes(self):
...
+
+# fmt: on
+class TestNIGLoss:
+    def test_main(self):
+        with pytest.raises(ValueError):
+            NIGLoss(reg_weight=-1)
+
+        loss = NIGLoss(reg_weight=1e-2)
+
+        loss(*torch.rand((1, 4)).split(1, dim=-1), torch.randn(1, 1))
```
Comment (o-laurent) [inline tests/test_losses.py:47]: Perhaps include one assert based on an expected result?

File: auto_tutorials_source/tutorial_der_cubic.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+"""
+Deep Evidential Regression Tutorial
+===================================
+
+This tutorial aims to provide an introductory overview of Deep Evidential Regression (DER) using a practical example. We will demonstrate the application of DER by tackling the problem of fitting :math:`y=x^3` using a Multi-Layer Perceptron (MLP) neural network model. The output layer of the MLP would have four outputs, and will be trained by minimizing the Normal Inverse-Gamma (NIG) loss function.
```
Comment (o-laurent) [inline auto_tutorials_source/tutorial_der_cubic.py:8]: Great thanks! I would just change to the present tense (I'll do it)

File: torch_uncertainty/routines/regression.py [hunk 5]
```diff
@@ -88,7 +117,13 @@ def training_step(
         inputs, targets = batch
         logits = self.forward(inputs)
 
-        if self.dist_estimation:
+        if self.dist_estimation == 4:
+            means, v, alpha, beta = logits.split(1, dim=-1)
```
Comment (o-laurent) [inline torch_uncertainty/routines/regression.py:121]: In the longer term, we have to think of some way to specify the distribution that we want to use, although it is true that distributions with four parameters per dim are likely to remain NIGs.

**Review by o-laurent — COMMENTED** [review]
Thanks for the nice work! I'll give some thoughts on the comments that I left and make some changes.

**Comment by badrmarani** [comment]
Bonjour @o-laurent!

Thank you for your kind feedback. I will certainly be checking your reviews and addressing any suggested changes as soon as possible. I'd like to let you know that I'm planning to make another pull request by the end of this week to add Monte-Carlo Dropout (#40).

On another note, I recently read your paper on Packed-Ensembles and find it very interesting. I noticed that the library currently has an implementation of the Conv2d layer. If it's okay with you, I will open an issue and adjust my schedule for next week to work on adding Conv1d and Conv3d layers.

I'm genuinely interested in contributing to this project, so if there are any other issues to resolve or papers to implement, please feel free to let me know. I'm more than willing to work on them.

Best,

File: tests/test_losses.py (hunk 3 above)
Comment (badrmarani) [reply tests/test_losses.py]: I've added the expected outcome when all four parameters and the target are set to 1.

File: torch_uncertainty/routines/regression.py (hunk 5 above)
Comment (badrmarani) [reply torch_uncertainty/routines/regression.py:121]: I think it would be better to temporarily assume that the `dist_estimation` argument is equal to the `num_outputs` argument and set them to be the same. What are your thoughts?

File: torch_uncertainty/datamodules/cubic_regression.py (hunk 2 above)
Comment (alafage) [reply torch_uncertainty/datamodules/cubic_regression.py:1]: I second the idea of implementing this as a dataset rather than a datamodule. After discussions with @o-laurent, we will add two new folders within `datasets/`: `classification/` and `regression/` to organize our datasets further. The cubic regression dataset would be included in the `datasets/regression/toy.py` file.

File: torch_uncertainty/routines/regression.py (hunk 5 above)
Comment (alafage) [reply torch_uncertainty/routines/regression.py:121]: I had the same kind of problem with another project of mine, I would suggest creating a function like this:

```py
def get_distribution(dist_name: str, dist_params: Tensor) -> Distribution:
    ...
```
I think it could be interesting to use the `Distribution` class from Pytorch, but it might be too much. A dictionary with the parameters should do it nicely too.

File: torch_uncertainty/datamodules/cubic_regression.py (hunk 2 above)
Comment (o-laurent) [reply torch_uncertainty/datamodules/cubic_regression.py:1]: For information, I've started writing the agreed modifications.

**Comment by o-laurent** [comment]
> On another note, I recently read your paper on Packed-Ensembles and find it very interesting. I noticed that the library currently has an implementation of the Conv2d layer. If it's okay with you, I will open an issue and adjust my schedule for next week to work on adding Conv1d and Conv3d layers.

Thank you so much for your interest :). We already have an implementation of Conv1d in another project (@alafage), so it may be easier for us to deploy it in this library, but thanks for the proposal.

> I'm genuinely interested in contributing to this project, so if there are any other issues to resolve or papers to implement, please feel free to let me know. I'm more than willing to work on them.

If you are interested in the project, I can invite you to join the GitHub team as a developer if you would like to. We have a lot of ideas to improve and extend the library, and I could detail some of them by starting a GitHub project. Do you have a specific interest in the field of uncertainty? If so, we could try to think about new ideas that match your own projects!

File: torch_uncertainty/routines/regression.py (hunk 5 above)
Comment (badrmarani) [reply torch_uncertainty/routines/regression.py:121]: Sure, I'll begin working on it. Do I need to open a separate issue for this, or can I work on it within this PR?

**Comment by o-laurent** [comment]
The PR now seems ready for merge. I'll do it tonight if there is no opposition.

**Comment by badrmarani** [comment]
Yes, I believe this PR is ready to be merged. Thank you, @o-laurent and @alafage, for your assistance. I truly appreciate it.

**Comment by badrmarani** [comment]
I'm definitely interested in this project and would love to join.

During my recent internship at Michelin, I worked on combining uncertainty quantification with deep generative models to help address the problem of generating high-quality and synthesizable molecules. I tested various models, including DER, and MC Dropout.

I have several papers in mind that I'd like to contribute to this library. These include [this paper](https://arxiv.org/pdf/2003.02037.pdf), [Wasserstein Dropout](https://arxiv.org/pdf/2012.12687.pdf), [MC-DropConnect](https://arxiv.org/pdf/1906.04569.pdf), and [MC-SoftDropConnect](https://arxiv.org/pdf/2201.08418.pdf).

**Comment by o-laurent** [comment]
Wow, it is a very nice project! 
They would all make excellent additions to the library, and I'd be happy to look into these methods with you!
I've sent you an invite to join the team.
